show export and import progress

manishbisht opened the pull request:
Fixes: #6311 and #12401 

Signed-off-by: Manish Bisht <manish.bisht490@gmail.com>

Before submitting pull request, please check that every commit:

- [x] Has proper Signed-Off-By
- [x] Has commit message which describes it
- [x] Is needed on it's own, if you have just minor fixes to previous commits, you can squash them
- [x] Any new functionality is covered by tests

manishbisht committed:
show export progress

Signed-off-by: Manish Bisht <manish.bisht490@gmail.com>

nijel commented on export.php:
```diff
@@ -538,3 +538,7 @@
 }
+
+$cfgRelation = PMA_getRelationsParam();
+$sql_query = 'DELETE FROM '.$cfgRelation['db'].'.' . $cfgRelation['progress'] . ' WHERE type = "export"';
```
You should properly escape database and table name here.

manishbisht replied:
It can be done using this right `PhpMyAdmin\Util::backquote($cfgRelation['db'])` ?

ibennetch replied:
Yes, `Util::backquote()` is what I would use here.

ibennetch replied:
Before running the query (here and at other points in the code), we should test if $cfgRelation['progress'] exists, perhaps with
```
if (!empty($cfgRelation['progress'])) {
    $sql_query = 'DELETE FROM '.$cfgRelation['db'].'.' . $cfgRelation['progress'] . ' WHERE type = "export"';
    PMA_queryAsControlUser($sql_query); 
}
```

ibennetch replied:
Before running the query (here and at other points in the code), we should test if $cfgRelation['progress'] exists, perhaps with
```
if (!empty($cfgRelation['progress'])) {
    $sql_query = 'DELETE FROM '.$cfgRelation['db'].'.' . $cfgRelation['progress'] . ' WHERE type = "export"';
    PMA_queryAsControlUser($sql_query); 
}
```

manishbisht replied:
To optimize this can we perform the sql query using ajax request on configuration database.
So that instead of running setInterval() always we can do something like this.
1. Start setInterval when the button is clicked.
2. Track regularly the progress percentage and once we reach 100 stop setInterval.
3. Perform the ajax request to delete the data.

nijel commented on libraries/export.lib.php:
```diff
@@ -606,7 +606,18 @@ function PMA_exportDatabase(
     foreach ($tables as $table) {
+        $count++;
+        array_push($data, $table);
+        $sql_query = 'UPDATE '.$cfgRelation['db'].'.' . $cfgRelation['progress'] . ' SET data = "' . implode(",", $data) . '", value = ' . $count . ' WHERE type = "export"';
```
* What happens if more users will be exporting at same time? You need to identify session as well or generate UUID for every upload. 
* Missing escaping around data leading to SQL injection vulnerability for control user.

manishbisht replied:
escaping data can be done using `$GLOBALS['dbi']->escapeString($count)` right ?
do we have any function available that generate uuid ?

ibennetch replied:
As far as I know, we don't have anything to make a UUID. There's a [bit of code](http://php.net/manual/en/function.uniqid.php#94959) that everyone seems to reference as being the way to do it in PHP, but without the author posting any license information it's not clear to me if we're able to use it. PHP doesn't yet have a UUID function itself, meaning I think the best solution is to use MySQL's [`UUID()`](https://dev.mysql.com/doc/refman/5.7/en/miscellaneous-functions.html#function_uuid) function. Perhaps Michal has a different suggestion, though.

ibennetch replied:
And yes, I'd use `escapeString` here.

nijel replied:
Using MySQL's function would certainly work. Another option is to use Session ID as identifier, that would limit to one upload per session, what is probably okay.

That reminds me that the table should also contain timestamp to allow cleanup of no longer needed entries...

manishbisht replied:
I have added a custom JavaScript function to generate uuid because to update the progress on page it is needed on JavaScript part also. Updated the PR. Please take a look into it.

nijel commented on js/export.js:
```diff
@@ -978,4 +978,6 @@ AJAX.registerOnload('export.js', function () {
     });
-});
+
+    setInterval(function(){ showProgress("export"); }, 1000);
```
This probably should happen after submitting export form, not on loading the export page....

manishbisht replied:
Yes, I was looking for this option only. This was not coming in my mind. Thanks :)

ibennetch commented:
When I open the main page, I get this error/warning message:

> Notice in ./libraries/relation.lib.php#562
>  Undefined index: progress
> 
> Backtrace
> 
> ./libraries/relation.lib.php#69: PMA_checkRelationsParam()
> ./libraries/user_preferences.lib.php#50: PMA_getRelationsParam()
> ./libraries/classes/Config.php#912: PMA_loadUserprefs()
> ./libraries/common.inc.php#862: PhpMyAdmin\Config->loadUserPreferences()
> ./index.php#22: require_once(./libraries/common.inc.php)

manishbisht commented:
This is due to some cookie issue. Try running in incognito mode or clear cookies before running.
Also have to re-imported the `.sql` file in configuration database ?

ibennetch commented:
So the reason for my error was because I hadn't yet configured the table in `config.inc.php`. We'll have to deal gracefully with situations where the user either hasn't configured it or the table doesn't exist.

I made a comment in-line with some code, but you can also see for instance libraries/classes/Bookmark.php:114 for an example of how we might deal with this. We have to test the array to make sure $cfgRelation['progress'] isn't empty first, before trying to use it. We initialize it as part of the default configuration to an empty value, so here we'll test if the user has configured it.

Also, you should add the new directive to `libraries/config.default.php` with a blank assignment such as `$cfg['Servers'][$i]['progress'] = '';`. You can add it right after the export_template area  around line 503.

We should also add this new directive to `libraries/relation.lib.php` around line 353; this file is used by chk_rel.php which warns the user if their table structure is outdated.

manishbisht commented:
Updated the PR making the suggested changes.

manishbisht commented:
It now support multiple user export using uuid. There is the change in the template of the table so delete the pma_progress from configuration database and re import the `.sql` file.

manishbisht commented:
The export progress is almost done. I am now waiting for the review/feedback if it need any changes.

manishbisht committed:
show import progress

Signed-off-by: Manish Bisht <manish.bisht490@gmail.com>

manishbisht commented:
progress is now showing for both import and export. Here is the output for both.
![screenshot from 2017-08-06 12-54-17](https://user-images.githubusercontent.com/14255990/29001374-adf2d750-7aa6-11e7-8ae2-47ee59e33e91.png)
![screenshot from 2017-08-06 12-54-46](https://user-images.githubusercontent.com/14255990/29001377-b4a17c96-7aa6-11e7-8ed1-9610a24f8ab4.png)

ibennetch commented on js/export.js:
```diff
@@ -7,8 +7,7 @@
  */
-function disable_dump_some_rows_sub_options()
-{
+function disable_dump_some_rows_sub_options() {
```
It's better to keep such style fixes as a separate commit so it doesn't look like this was part of your code for the progress bar. Unstaging it from your commits might be too much hassle now, though. They're good fixes to perform when working with these older library files, but from the aspect of maintaining efficient history, it's best as another commit.

ibennetch commented:
It works pretty well for me, except on export the whole status box only fills with information after the download completes.

![image](https://user-images.githubusercontent.com/1443710/29045702-85c23cd4-7b92-11e7-8856-3ae888f3e895.png)

This is shown after the download finishes; while download is in progress it just shows a small bar like this:
![screen shot 2017-08-07 at 5 07 18 pm](https://user-images.githubusercontent.com/1443710/29045824-f055b9b8-7b92-11e7-90fe-4a4b53e509ee.png)

Any idea about why I'm seeing that?

ibennetch commented:
Also, I'm having trouble getting it to work on Firefox. Chrome works okay, but Firefox gives me an error code 0:

![ff](https://user-images.githubusercontent.com/1443710/29046358-343e18f8-7b95-11e7-826f-4d460ac632fa.png)

Does anyone else see that error?

manishbisht committed:
export progress bug fixed

Signed-off-by: Manish Bisht <manish.bisht490@gmail.com>

manishbisht commented:
I have added a little improvement for export progress and removed the progress bar because the export file is downloaded and in parallel it blocks the set Interval and Set Timeout till the download completed.

I am seeing the Mozilla error on the master branch also.
